File: iosApp/Models/EditParameters.swift
import Foundation
import CoreGraphics

/// Core data structure holding all editing parameters.
struct EditParameters: Codable, Equatable {
    var exposure: Float = 0
    var contrast: Float = 0
    var highlights: Float = 0
    var shadows: Float = 0
    var saturation: Float = 0
    var vibrance: Float = 0
    var warmth: Float = 0
    var sharpness: Float = 0
    var texture: Float = 0
    var clarity: Float = 0
    var dehaze: Float = 0

    var cropRect: CodableCGRect?
    var rotationCount: Int = 0

    static let `default` = EditParameters()

    var isDefault: Bool {
        return self == EditParameters.default
    }
}

struct CodableCGRect: Codable, Equatable {
    var x: CGFloat
    var y: CGFloat
    var width: CGFloat
    var height: CGFloat

    var cgRect: CGRect {
        CGRect(x: x, y: y, width: width, height: height)
    }

    init(_ rect: CGRect) {
        self.x = rect.origin.x
        self.y = rect.origin.y
        self.width = rect.size.width
        self.height = rect.size.height
    }
}

File: shared/src/commonMain/kotlin/com/example/photoeditor/EditParameters.kt
package com.example.photoeditor

import kotlinx.serialization.Serializable

@Serializable
data class CropRect(
    val x: Float,
    val y: Float,
    val width: Float,
    val height: Float
)

@Serializable
data class EditParameters(
    val exposure: Float = 0f,      // -100 ~ +100
    val contrast: Float = 0f,
    val highlights: Float = 0f,
    val shadows: Float = 0f,
    val saturation: Float = 0f,
    val vibrance: Float = 0f,
    val warmth: Float = 0f,
    val sharpness: Float = 0f,
    val texture: Float = 0f,
    val clarity: Float = 0f,
    val dehaze: Float = 0f,
    val cropRect: CropRect? = null,
    val rotationCount: Int = 0     // 0-3, clockwise 90° count
) {
    val isDefault: Boolean get() = this == EditParameters()
}

/** Returns a new EditParameters with rotationCount incremented by 1 (mod 4 semantics) */
fun EditParameters.rotate(): EditParameters = copy(rotationCount = (rotationCount + 1) % 4)
